packages/patrol/lib/src/custom_finders/patrol_tester.dart

```diff
@@ -401,56 +401,71 @@ class PatrolTester {
+  ///  * scrolls until until [finder] finds at least one *existing* widget
   ///
   ///  * waits until [view] is visible
   ///
-  ///  * uses [WidgetController.firstElement] instead of
-  ///    [WidgetController.element], which avoids [StateError] being thrown in
-  ///    situations when [finder] finds more than 1 visible widget
+  ///  * if the [view] finder finds more than 1 widget, it scrolls the first one
+  ///    instead of throwing a [StateError]
+  ///
+  ///  * if the [finder] finder finds more than 1 widget, it scrolls to the
+  ///    first one instead of throwing a [StateError]
+  ///
+  ///  * can drag any widget, not only a [Scrollable]
+  ///
+  ///  * performed drag is slower (it takes some time to performe dragging
+  ///    gesture, half a second by default)
+  ///
+  ///  * you can configure, which version of pumping is performed between
+  ///    each drag gesture ([pump], [pumpAndSettle] or [pumpAndTrySettle]).
   ///
   /// See also:
   ///  * [PatrolTester.config.andSettle], which controls the default behavior if
   ///    [andSettle] is null
+  ///  * [PatrolTester.dragUntilVisible], which scrolls to visible widget,
+  ///    not only existing one.
   Future<PatrolFinder> dragUntilExists({
     required Finder finder,
     required Finder view,
     required Offset moveStep,
     int maxIteration = defaultScrollMaxIteration,
-    Duration duration = const Duration(milliseconds: 50),
+    Duration pumpDuration = const Duration(milliseconds: 50),
+    Duration dragDuration = const Duration(milliseconds: 500),
     @Deprecated('Use settleBehavior argument instead') bool? andSettle,
     SettlePolicy? settlePolicy,
   }) {
     return TestAsyncUtils.guard(() async {
-      final viewPatrolFinder = PatrolFinder(finder: view, tester: this);
+      var viewPatrolFinder = PatrolFinder(finder: view, tester: this);
       await viewPatrolFinder.waitUntilVisible();
+      viewPatrolFinder = viewPatrolFinder.hitTestable().first;
+      final settle = chooseSettlePolicy(
+        andSettle: andSettle,
+        settlePolicy: settlePolicy,
+      );
 
       var iterationsLeft = maxIteration;
       while (iterationsLeft > 0 && finder.evaluate().isEmpty) {
-        await tester.drag(view, moveStep);
-        await tester.pump(duration);
+        await tester.timedDrag(viewPatrolFinder, moveStep, dragDuration);
+        await _performPump(
+          settlePolicy: settle,
+          settleTimeout: config.settleTimeout,
+        );
         iterationsLeft -= 1;
       }
-      await Scrollable.ensureVisible(tester.firstElement(finder));
 
-      final settle = chooseSettlePolicy(
-        andSettle: andSettle,
-        settlePolicy: settlePolicy,
-      );
-      await _performPump(
-        settlePolicy: settle,
-        settleTimeout: config.settleTimeout,
-      );
+      if (iterationsLeft <= 0) {
+        throw WaitUntilExistsTimeoutException(
+          finder: finder,
+          // TODO: set reasonable duration
```

Comment:
👀

Reply:
It's because I couldn't figure out what we should log here, bc we weren't waiting explicitly. And I still don't know what to put here.

Maybe we should implement separate exception for this situation.